pkg/capi/gcp.go

```diff
@@ -0,0 +1,182 @@
+package capi
+
+import (
+	"context"
+	"fmt"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	configv1 "github.com/openshift/api/config/v1"
+	mapiv1 "github.com/openshift/api/machine/v1beta1"
+	framework "github.com/openshift/cluster-api-actuator-pkg/pkg/framework"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	gcpv1 "sigs.k8s.io/cluster-api-provider-gcp/api/v1beta1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/envtest/komega"
+	yaml "sigs.k8s.io/yaml"
+)
+
+const (
+	infraAPIVersion        = "infrastructure.cluster.x-k8s.io/v1beta1"
+	gcpMachineTemplateName = "gcp-machine-template"
+)
+
+var (
+	ctx         = context.Background()
+	clusterName string
+	cl          client.Client
+)
+
+var _ = Describe("Cluster API GCP MachineSet", Ordered, func() {
+	var gcpMachineTemplate *gcpv1.GCPMachineTemplate
+	var machineSet *clusterv1.MachineSet
+	var mapiMachineSpec *mapiv1.GCPMachineProviderSpec
+	var ctx context.Context
+	var platform configv1.PlatformType
+	var clusterName string
+	var err error
+
+	BeforeAll(func() {
+		cl, err = framework.LoadClient()
+		Expect(err).NotTo(HaveOccurred(), "Failed to create Kubernetes client for test")
+		komega.SetClient(cl)
+		ctx = framework.GetContext()
+		platform, err = framework.GetPlatform(ctx, cl)
+		Expect(err).ToNot(HaveOccurred(), "Failed to get platform")
+		if platform != configv1.GCPPlatformType {
+			Skip("Skipping GCP E2E tests")
+		}
+		oc, _ := framework.NewCLI()
+		framework.SkipIfNotTechPreviewNoUpgrade(oc, cl)
+
+		infra, err := framework.GetInfrastructure(ctx, cl)
+		Expect(err).NotTo(HaveOccurred(), "Failed to get cluster infrastructure object")
+		Expect(infra.Status.InfrastructureName).ShouldNot(BeEmpty(), "infrastructure name was empty on Infrastructure.Status.")
+		clusterName = infra.Status.InfrastructureName
+
+		framework.CreateCoreCluster(ctx, cl, clusterName, "GCPCluster")
+		mapiMachineSpec = getGCPMAPIProviderSpec(cl)
+	})
+	AfterEach(func() {
+		if platform != configv1.GCPPlatformType {
+			// Because AfterEach always runs, even when tests are skipped, we have to
+			// explicitly skip it here for other platforms.
+			Skip("Skipping GCP E2E tests")
+		}
+		framework.DeleteCAPIMachineSets(ctx, cl, machineSet)
+		framework.WaitForCAPIMachineSetsDeleted(ctx, cl, machineSet)
+		framework.DeleteObjects(ctx, cl, gcpMachineTemplate)
+	})
+
+	diskTypes := map[string]gcpv1.DiskType{
+		"pd-standard": gcpv1.PdStandardDiskType,
+		"pd-ssd":      gcpv1.PdSsdDiskType,
+		//	"local-ssd":   gcpv1.LocalSsdDiskType,//boot disk cannot be local disk hence commenting it
```

Comment:
@damdo @RadekManak , do we want to have a webhook like in `mapi`

`error: machinesets.machine.openshift.io "miyadav-1811g-vk7vb-worker-f" could not be patched: admission webhook "validation.machineset.machine.openshift.io" denied the request: providerSpec.disks[0].type: Unsupported value: "local-ssd": supported values: "hyperdisk-balanced", "pd-balanced", "pd-ssd", "pd-standard"
`